[Build] Split Kernels into Separate `vllm-kernels` package

xli opened the pull request:
Initialize vllm-kernels package with a small refactoring in cmake files.

## Purpose

This is first step towards implementing #17419.

As splitting kernels into a new package involves changes to CI, dev workflow and release new package process, I will break all changes into multiple steps to minimize the impact.

This PR is the first step. 
In this PR, we initialize a new vllm-kernels package to package all c-extensions. This allows us to start to have a new vllm-kernels package without breaking any existing workflow, CI or ongoing changes.
One hook I have to put in vllm package is dynamic importing vllm_kernels modules as vllm's module, make them looks like same with all extensions are included in the vllm package. It makes sure the behavior are same for the following 2 cases:
1. Build kernels by vllm/setup.py and packaged inside the vllm package.
2. Build vllm-kernels package and vllm package without kernels (by specifying VLLM_NO_EXTENSION=1 in building process)
This way we won't break existing workflow and vllm package before everything is fully ready for splitting out vllm-kernels package.

Also added a simple integration test to prevent breaking change when vllm package / kernels compiling logic changed before we fully migrate over.

Moving forward plan:
1. add CI build to release new vllm-kernels package.
2. setup nightly build for vllm-kernels package
3. switch VLLM_USE_PRECOMPILED to use vllm-kernels nightly build and update vllm package to only build and package python code.
4. Move kernels tests and interfaces (e.g. _custom_ops.py, open for discussion) into vllm-kernels.
5. Switch to build vllm package with python / engine code only. Move cmake and csrs into kernels dir. Update CI to test vllm + vllm-kernels packages.
6. Clean up vllm-kernels module dynamic import aliases, direct import vllm_kernels modules in vllm.

## Test Plan and Result

Be able to serve llama3 8b model, and response a test request

Test cases:
1. build and install wheels for vllm-kernels and vllm (with VLLM_NO_EXTENSION=1)
2. install editable vllm-kernels and vllm (with VLLM_NO_EXTENSION=1)
3. install editable vllm pacakge with VLLM_TARGET_DEVICE=cuda
4. install vllm-kernels and install editable vllm pacakge with VLLM_TARGET_DEVICE=cuda: in this case it will use vllm-kernels modules. We don't expect this env setup, kept code simple when importing vllm-kernels modules into vllm module. 

Build scripts see kernels/Makefile targets:
* wheel: build vllm-kernels wheel
* install: install editable vllm-kernels
* vllm-wheel: build vllm wheel without kernels
* vllm-install: install editable vllm without kernels

## Test Result

All test cases results are good

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
</details>

xli committed:
[refactoring] add BUILD_DESTINATION for customizing cmake build destination in setup.py

Signed-off-by: Xiao Li <ilx@meta.com>

xli committed:
Initialize new vllm-kernels package

Signed-off-by: Xiao Li <ilx@meta.com>

NickLucche commented:
This is awesome, thanks for your work!
This will eventually also enable oot hardware to provide their own set of kernels to support vllm.

dtrifiro commented on kernels/setup_cuda_env.sh:
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
```
Do we need this file?

xli replied:
I will remove this one, it's my local file, added accidentally

xli commented on cmake/external_projects/flashmla.cmake:
```diff
@@ -1,6 +1,6 @@
 include(FetchContent)
 
-# If FLASH_MLA_SRC_DIR is set, flash-mla is installed from that directory 
```
removed these changes

xli committed:
add an integration test for validating vllm-kernels package works with vllm without kernels package

Signed-off-by: Xiao Li <ilx@meta.com>

xli committed:
run new integration test in test pipeline

Signed-off-by: Xiao Li <ilx@meta.com>

xli committed:
add unit test for vllm/kernels _alias_modules; and fix a log bug in _alias_modules

Signed-off-by: Xiao Li <ilx@meta.com>

xli committed:
Fix build on amd

Rebased to trunk healthy revision, changed kernels torch version to match
recent change (2.8.0);
Removed assertion & eval configs that don't work on amd.
Added VLLM_NO_EXTENSION option to vllm setup.py for building package without
extension, so that device specific requirements are required by the package.

Signed-off-by: Xiao Li <ilx@meta.com>

xli committed:
update makefile script

Signed-off-by: Xiao Li <ilx@meta.com>

xli committed:
fix unit test after renamed the test file name; fix pre-commit verification

Signed-off-by: Xiao Li <ilx@meta.com>

xli committed:
fix type checking in test

Signed-off-by: Xiao Li <ilx@meta.com>

xli committed:
limit max jobs and nvcc threads for building c exts in test

Signed-off-by: Xiao Li <ilx@meta.com>
